### ISSUE TITLE
SELinux / pam_namespace won't let TigerVNC start

### ISSUE THREAD

**Issue by JackTatBoeing** [TigerVNC/tigervnc#1568]
I'm running Oracle Linux 7.9 with current updates. I see the latest version of TigerVNC available for this is 1.8.x. However, I'm hardening a system, and I enabled SELinux and set FIPS mode. Things worked fine (I think) until I got a Nessus scan back that TigerVNC had to be at version 1.10 or higher. I looked on rpmfind and found a version of TigerVNC 1.12.x listed as being compiled for el7. I stopped and disabled all instances of vncserver. I downloaded all the modules, removed the old version, then installed:
tigervnc-1.12.90-1.el7.x86_64.rpm
tigervnc-icons-1.12.90-1.el7.noarch.rpm
tigervnc-license-1.12.90-1.el7.noarch.rpm
tigervnc-selinux-1.12.90-1.el7.noarch.rpm
tigervnc-server-1.12.90-1.el7.x86_64.rpm
tigervnc-server-minimal-1.12.90-1.el7.x86_64.rpm
tigervnc-server-module-1.12.90-1.el7.x86_64.rpm

I deleted all the old vnc stuff, like `/etc/systemd/system/vncserver@:1.service`. I saw an SELinux post on here so I deleted the original `~/.vnc` directories for the two users. I `su -` to those two users and ran `vncpasswd` to recreate their passwords. I removed all the directories in `/tmp/.X11-users`. I updated the `/etc/tigervnc/vncserver.users` file with the to users:
```
1:=stssvr
9:=admin
```

I ran `systemctl daemon-reload`. I tried starting the first vncserver by running, `systemctl unmask vncserver@:1`, `systemctl enable vncserver@:1`, and `systemctl start vncserver@:1`. This command returns:
```
Job for vncserver@:1.service failed because the control process exited with error code. See "systemctl status vncserver@:1.service" and "journalctl -xe" for details.
```

`systemctl status vncserver@:1` returns:
```
● vncserver@:1.service - Remote desktop service (VNC)
   Loaded: loaded (/etc/systemd/system/vncserver@:1.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Thu 2022-12-29 22:08:30 UTC; 30s ago
  Process: 13646 ExecStart=/usr/libexec/vncsession-start %i (code=exited, status=71)

Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: Starting Remote desktop service (VNC)...
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 vncsession[13653]: pam_namespace(tigervnc:session): Unable to unshare from parent namespace, Operation not permitted
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 vncsession-start[13646]: Failed to start session
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: vncserver@:1.service: control process exited, code=exited status=71
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: Failed to start Remote desktop service (VNC).
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: Unit vncserver@:1.service entered failed state.
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: vncserver@:1.service failed.
```

`journalctl -xe` shows:
```
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 kernel: type=1400 audit(1672351710.782:652): avc:  denied  { create } for  pid=1401 comm="systemd-logind" name="us
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 kernel: type=1300 audit(1672351710.782:652): arch=c000003e syscall=83 success=no exit=-13 a0=563ea1d55a47 a1=1ed a
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 kernel: type=1327 audit(1672351710.782:652): proctitle="/usr/lib/systemd/systemd-logind"
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd-logind[1401]: Failed to create /run/user: No such file or directory
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 kernel: type=1105 audit(1672351710.783:653): pid=13653 uid=0 auid=1101 ses=33 subj=system_u:system_r:vnc_session_t
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 vncsession[13653]: pam_systemd(tigervnc:session): Failed to create session: No such file or directory
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 vncsession[13653]: pam_unix(tigervnc:session): session opened for user stssvr by (uid=0)
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 vncsession-start[13646]: Failed to start session
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 vncsession[13653]: pam_open_session failed: 14 (Cannot make/remove an entry for the specified session)
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: vncserver@:1.service: control process exited, code=exited status=71
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: Failed to start Remote desktop service (VNC).
-- Subject: Unit vncserver@:1.service has failed
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit vncserver@:1.service has failed.
--
-- The result is failed.
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: Unit vncserver@:1.service entered failed state.
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: vncserver@:1.service failed.
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 kernel: type=1130 audit(1672351710.796:654): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:ini
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 polkitd[1399]: Unregistered Authentication Agent for unix-process:13639:514882 (system bus name :1.160, object pat
Dec 29 22:09:01 atcsvr-fips-uefi-12-28-22 systemd[1]: Configuration file /etc/systemd/system/vncserver@:1.service is marked world-inaccessible. This has no
Dec 29 22:09:27 atcsvr-fips-uefi-12-28-22 sshd[13089]: Timeout, client not responding.
Dec 29 22:09:27 atcsvr-fips-uefi-12-28-22 sshd[13089]: pam_unix(sshd:session): session closed for user root
Dec 29 22:09:27 atcsvr-fips-uefi-12-28-22 kernel: type=1106 audit(1672351767.021:655): pid=13089 uid=0 auid=0 ses=29 subj=unconfined_u:unconfined_r:unconfin
Dec 29 22:09:27 atcsvr-fips-uefi-12-28-22 kernel: type=1104 audit(1672351767.034:656): pid=13089 uid=0 auid=0 ses=29 subj=unconfined_u:unconfined_r:unconfin
Dec 29 22:09:27 atcsvr-fips-uefi-12-28-22 kernel: type=1106 audit(1672351767.046:657): pid=13089 uid=0 auid=0 ses=29 subj=unconfined_u:unconfined_r:unconfin
Dec 29 22:09:27 atcsvr-fips-uefi-12-28-22 kernel: type=1113 audit(1672351767.058:658): pid=13089 uid=0 auid=0 ses=29 subj=unconfined_u:unconfined_r:unconfin
Dec 29 22:09:27 atcsvr-fips-uefi-12-28-22 kernel: type=2404 audit(1672351767.058:659): pid=13089 uid=0 auid=0 ses=29 subj=unconfined_u:unconfined_r:unconfin
Dec 29 22:09:27 atcsvr-fips-uefi-12-28-22 kernel: type=2404 audit(1672351767.058:660): pid=13089 uid=0 auid=0 ses=29 subj=unconfined_u:unconfined_r:unconfin
Dec 29 22:09:27 atcsvr-fips-uefi-12-28-22 kernel: type=2404 audit(1672351767.058:661): pid=13089 uid=0 auid=0 ses=29 subj=unconfined_u:unconfined_r:unconfin
```

There were some earlier messages about pam_namespace problems, so I read a bit about pam_namespace so I looked at the `/etc/pam.d/vncserver` file:
```
#%PAM-1.0
# pam_selinux.so close should be the first session rule
-session   required     pam_selinux.so close
session    required     pam_loginuid.so
-session   required     pam_selinux.so open
session    required     pam_namespace.so
session    optional     pam_keyinit.so force revoke
session    required     pam_limits.so
-session   optional     pam_systemd.so
session    required     pam_unix.so
-session   optional     pam_reauthorize.so prepare
```

The pam_namespace line seemed correct so I didn't make any changes. q

Here is my `/etc/selinux/conf` file:
```
# This file controls the state of SELinux on the system.
# SELINUX= can take one of these three values:
#     enforcing - SELinux security policy is enforced.
#     permissive - SELinux prints warnings instead of enforcing.
#     disabled - No SELinux policy is loaded.

# Per CCE-27334-2: Set SELINUX=enforcing in /etc/sysconfig/selinux
SELINUX=enforcing

# SELINUXTYPE= can take one of three values:
#     targeted - Targeted processes are protected,
#     minimum - Modification of targeted policy. Only selected processes are protected.
#     mls - Multi Level Security protection.
SELINUXTYPE=targeted
```

FIPS is enabled on this system. 

It would appear to me that, perhaps, SELinux is interfering with TigerVNC starting. I've done some searching on the Internet for things related to this, but there isn't much.

Is this something you could help with? Do you need more information?

Thanks, Jack

**Comment by CendioOssman** [comment]
I tried your setup on a RHEL 7 system here without issues, so there seems to be something specific to your system.

> I downloaded all the modules, removed the old version, then installed:
> tigervnc-1.12.90-1.el7.x86_64.rpm
> ...

Where, exactly, did you get those packages? The version number suggests our beta, but I want to make sure it isn't some modified build.

> There were some earlier messages about pam_namespace problems,

pam_namespace is not enabled by default. What do you have in `/etc/security/namespace.conf` and `/etc/security/namespace.d`?

Exactly what have you done as part of hardening this system? Can you see the problems on a freshly installed system?

**Comment by JackTatBoeing** [comment]
Hi, and thanks for looking at this. The files I used were from:
tigervnct1.12.90-1.el7.x86_64.rpm					https://rpmfind.net/linux/rpm2html/search.php?query=tigervnc
tigervnc-icons-1.12.90-1.el7.noarch.rpm			https://rpmfind.net/linux/rpm2html/search.php?query=tigervnc-icons
tigervnc-license-1.12.90-1.el7.noarch.rpm			https://rpmfind.net/linux/rpm2html/search.php?query=tigervnc-license
tigervnc-selinux-1.12.90-1.el7.noarch.rpm			https://rpmfind.net/linux/rpm2html/search.php?query=tigervnc-selinux
tigervnc-server-1.12.90-1.el7.x86_64.rpm			https://rpmfind.net/linux/rpm2html/search.php?query=tigervnc-server
tigervnc-server-minimal-1.12.90-1.el7.x86_64.rpm	https://rpmfind.net/linux/rpm2html/search.php?query=tigervnc-server-minimal
tigervnc-server-module-1.12.90-1.el7.x86_64.rpm		https://rpmfind.net/linux/rpm2html/search.php?query=tigervnc-server-module

As far as hardening, not really all that much. I enabled FIPS 140 mode -- I ran a script to do this (attached). I removed all versions of Java except Oracle's version 1.8. I updated QEMU to version 4.2.1-22. I enabled SELinux and, created /.autorelabel, and ran 'fixfiles onboot' before rebooting. I removed a java directory from the SmartSVN directory. I removed all instances of log4j.jar. 

And today I note the pam_namespace error is back:
● vncserver@:1.service - Remote desktop service (VNC)
   Loaded: loaded (/etc/systemd/system/vncserver@:1.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Thu 2022-12-29 22:08:30 UTC; 15h ago
  Process: 13646 ExecStart=/usr/libexec/vncsession-start %i (code=exited, status=71)

Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: Starting Remote desktop service (VNC)...
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 vncsession[13653]: pam_namespace(tigervnc:session): Unable to unshare from parent namespace, Operation n...rmitted
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 vncsession-start[13646]: Failed to start session
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: vncserver@:1.service: control process exited, code=exited status=71
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: Failed to start Remote desktop service (VNC).
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: Unit vncserver@:1.service entered failed state.
Dec 29 22:08:30 atcsvr-fips-uefi-12-28-22 systemd[1]: vncserver@:1.service failed.
Hint: Some lines were ellipsized, use -l to show in full.

I've included the script I used to enable FIPS (renamed from .sh to .txt). I've included the namespace.conf file you asked about. There is nothing in the /etc/security/namespace.d folder. 

[enable_fips.txt](https://github.com/TigerVNC/tigervnc/files/10325024/enable_fips.txt)
[namespace.conf.txt](https://github.com/TigerVNC/tigervnc/files/10325040/namespace.conf.txt)

Here are the two main vncserver config files:
[vncserver.users.txt](https://github.com/TigerVNC/tigervnc/files/10325052/vncserver.users.txt)
[vncserver@%3A1.service.txt](https://github.com/TigerVNC/tigervnc/files/10325054/vncserver%40.3A1.service.txt)

I'm not sure what else might be helpful, but thanks once again. 
Jack

**Comment by JackTatBoeing** [comment]
Hi again,
I uncommented the @home entry in the namespace config file, and then I couldn't log in at all, locally or remotely. Or even in emergency mode. So... I had to start over. I was very careful recreating the hardening I did earlier. Thus far I was able to remove the 1.8 Tigervnc and install 1.12 Tigervnc and it is still working. I'm also taking relatively frequent backups to provide points of recovery.

Anyway, thanks for your help, but I don't want you to keep working on something that is no longer a problem -- for the time being, at least. :)
Jack
